src/template.c

```diff
@@ -5958,22 +5958,23 @@ void TemplateInstance::semantic(Scope *sc, Expressions *fargs)
         this->tnext = inst->tnext;
         inst->tnext = this;
 
-        // If the first instantiation was in speculative context, but this is not:
-        if (tinst && !inst->tinst && !inst->minst)
+        if (minst && minst->isRoot() && !(inst->minst && inst->minst->isRoot()))
         {
-            // Reconnect the chain if this instantiation is not in speculative context.
+            // Swap the minst and tinst between inst and this.
+            Module *mi = minst;
             TemplateInstance *ti = tinst;
-            while (ti && ti != inst)
-                ti = ti->tinst;
-            if (ti != inst)    // Bugzilla 13379: Prevent circular chain
-                inst->tinst = tinst;
-        }
+            minst = inst->minst;
+            tinst = inst->tinst;
+            inst->minst = mi;
+            inst->tinst = ti;
```

Comment:
I updated to make the comment detailed more. Is that enough?

A poor solution is adding one bool flag `bool TemplateInstance::isAlreadyInstantiatedInRoot;` and test&check it here. But it will increment the size of TemplateInstance object. As far as possible I'd like to save memory footprint of dmd.